src/main/java/org/owasp/esapi/codecs/JSONCodec.java

```diff
@@ -0,0 +1,242 @@
+     */
+    public String encodeCharacter( char[] immune, int codePoint )
+        throws IllegalArgumentException {
+
+        // Per the docs for HTMLEntityCodec: "WARNING: This method will silently discard
+        // invalid code points per the call to Character.isValidCodePoint( int ) method.
+        // WARNING!! Character based Codecs will silently transform code points that are
+        // not legal UTF code points into garbage data as they will cast them to chars.
+        if ( Character.isValidCodePoint( codePoint ) == false ) {
+            // throw new IllegalArgumentException( "Invalid codepoint '" + String.format("0x%04X", codePoint) + "'." );
+            return "";
+        }
+
+        if ( immune != null ) {
+            // More efficient than sort and binary search. If the immune array
+            // was presorted, then this could be O(log n). But we can't add the
+            // precondition now. It is too late in the game.
+            for ( Character ch : immune ) {
+                if ( charToCodepoint( ch ) == codePoint ) {
+                    return new String(Character.toChars(codePoint));
+                }
+            }
+        }
+
+        // Per the RFC... Two-character sequence escape representations of some
+        // popular characters
+        switch ( codePoint ) {
+            case '\b': return "\\b";
+            case '\f': return "\\f";
+            case '\r': return "\\r";
+            case '\n': return "\\n";
+            case '\t': return "\\t";
+            case '"':  return "\\\"";
+            case '/':  return  "\\/";
+            case '\\': return "\\\\";
+        }
+
+        // Per the RFC... All Unicode characters may be placed within the
+        // quotation marks, except for the characters that MUST be escaped:
+        // quotation mark, reverse solidus, and the control characters
+        // (U+0000 through U+001F).
+        if ( codePoint <=  0x1f ) {
+
+            return String.format("\\u%04x", codePoint);
+        }
+
+        return new String(Character.toChars(codePoint));
+    }
+
+
+    /**
+     * {@inheritDoc}
+     *
+     * Decodes special characters in encoded JSON strings.
+     *
+     * decodeCharacter will decode the encoded character sequences for popular characters
+     * Backspace (\b), Form Feed (\f), Carriage Return (\r), Line Feed (\n), Tab (\t),
+     * Double Quote ("), Forward slash (/) and Backslash (\). The function will also decode
+     * six-character sequences of \u0000 - \uffff. If the character is not encoded then a
+     * null character is returned.
+     *
+     * @param input a character sequence to decode
+     * @return the decoded version of the encoded character starting at index,
+     *     or null otherwise
+     *
+     * @throws IllegalArgumentException
+     *     if an invalid character sequence is encountered
+     */
+    public Integer decodeCharacter( PushbackSequence<Integer> input )
```

Comment:
The logic flow of this method feels strange to me.

First thing catching my eye is the use of the IllegalArgumentExeption on line 175 to jump to line 209.  I think the try/catch block should be refactored into an if statement.

I appreciate the value of the fast-return on 165, but I have concerns about the other 10 branches that can be used to escape this method.  
 - 8 for each case in the switch
  - 1 at the end for an IllegalArgumentException
  - 1 and the last on line 201.   
   
  If it is possible to reduce this to 3 exit cases I think it will be easier to maintain: 
  - 1 : fast return
  - 2 : Exception case
  - 3 : standard return

I may be able to offer a snippet this evening to assist with this, if that would be helpful.

Reply:
I have it down to 4 cases, which may be the minimum:
 1 - First character is null (fast return/no Exception)
 2 - First character is not null, Second character is Null (fast return w/Exception)
 3 - First & Second not null, but no decode can be performed (Exception)
 4 - First & Second not null, decode performed and returned (normal response)

All tests are passing and I'll push to your branch shortly.  I appreciate the consideration and opportunity to contribute to the effort.

Reply:
The updated `decodeCharacter` looks great. Awesome job.